src/number_theory/legendre_symbol/quadratic_char.lean

```diff
@@ -0,0 +1,331 @@
+Authors: Michael Stoll
+-/
+import tactic.basic
+import field_theory.finite.basic
+
+/-!
+# Quadratic characters of finite fields
+
+This file defines the quadratic character on a finite field `F` and proves
+some basic statements about it.
+
+## Tags
+
+quadratic character
+-/
+
+/-!
+### Some general results on finite fields
+-/
+
+section general
+
+/-- If `ring_char R = 2`, where `R` is a finite reduced commutative ring,
+then every `a : R` is a square. -/
+lemma is_square_of_char_two' {R : Type*} [fintype R] [comm_ring R] [is_reduced R] [char_p R 2]
+ (a : R) : is_square a :=
+exists_imp_exists (λ b h, pow_two b ▸ eq.symm h) $
+  ((fintype.bijective_iff_injective_and_card _).mpr ⟨frobenius_inj R 2, rfl⟩).surjective a
+
+namespace finite_field
+
+variables {F : Type*} [field F] [fintype F]
+
+/-- In a finite field of characteristic `2`, all elements are squares. -/
+lemma is_square_of_char_two (hF : ring_char F = 2) (a : F) : is_square a :=
+begin
+  haveI hF' : char_p F 2 := ring_char.of_eq hF,
+  exact is_square_of_char_two' a,
+end
+
+/-- If the finite field `F` has characteristic `≠ 2`, then it has odd cardinatlity. -/
+lemma odd_card_of_char_ne_two (hF : ring_char F ≠ 2) : fintype.card F % 2 = 1 :=
+begin
+  rcases finite_field.card F (ring_char F) with ⟨ n, hp, h ⟩,
+  have h₁ : odd ((ring_char F) ^ (n : ℕ)) :=
+  odd.pow ((or_iff_right hF).mp (nat.prime.eq_two_or_odd' hp)),
+  rwa [← h, nat.odd_iff] at h₁,
+end
+
+/-- Characteristic `≠ 2` implies that `-1 ≠ 1`. -/
+lemma neg_one_ne_one_of_char_ne_two (hF : ring_char F ≠ 2) : (-1 : F) ≠ 1 :=
+begin
+  have hc := char_p.char_is_prime F (ring_char F),
+  haveI hF' : fact (2 < ring_char F) := ⟨ lt_of_le_of_ne (nat.prime.two_le hc) (ne.symm hF) ⟩,
+  exact char_p.neg_one_ne_one _ (ring_char F),
+end
+
+/-- If `F` has odd characteristic, then for nonzero `a : F`, we have that `a ^ (#F / 2) = ±1`. -/
+lemma pow_dichotomy (hF : ring_char F ≠ 2) {a : F} (ha : a ≠ 0) :
+  a^(fintype.card F / 2) = 1 ∨ a^(fintype.card F / 2) = -1 :=
+begin
+  have h₁ := finite_field.pow_card_sub_one_eq_one a ha,
+  set q := fintype.card F with hq,
+  have hq : q % 2 = 1 := finite_field.odd_card_of_char_ne_two hF,
+  have h₂ := nat.two_mul_odd_div_two hq,
+  rw [← h₂, mul_comm, pow_mul, pow_two] at h₁,
+  exact mul_self_eq_one_iff.mp h₁,
+end
+
+/-- A unit `a` of a finite field `F` of odd characteristic is a square
+if and only if `a ^ (#F / 2) = 1`. -/
+lemma unit_is_sqare_iff (hF : ring_char F ≠ 2) (a : Fˣ) :
+  is_square a ↔ a ^ (fintype.card F / 2) = 1 :=
+begin
+  classical,
+  obtain ⟨g, hg⟩ := is_cyclic.exists_generator Fˣ,
+  obtain ⟨n, hn⟩ : a ∈ submonoid.powers g, { rw mem_powers_iff_mem_zpowers, apply hg },
+  have hodd := nat.two_mul_odd_div_two (finite_field.odd_card_of_char_ne_two hF),
+  split,
+  { rintro ⟨y, rfl⟩,
+    rw [← pow_two, ← pow_mul, hodd],
+    apply_fun (@coe Fˣ F _),
+    push_cast,
+    exact finite_field.pow_card_sub_one_eq_one (y : F) (units.ne_zero y),
+    exact units.ext, },
```

Comment:
This suggest that there are multiple goals open at the same time. Please use more `{}`.

Reply:
OK. (This is code I had copied over from the original `quadratic_reciprocity.lean` file.)

Reply:
Or use [`exacts [ ..., ... ]`](https://leanprover-community.github.io/mathlib_docs/tactics.html#exacts) but that would probably make the line too long here.